BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c

```diff
@@ -380,6 +382,117 @@ UpdateMadt (
   return EFI_SUCCESS;
 }
 
+/**
+  This function updates FWST ACPI data structure based on ESRT table.
+
+  @param[in]  Current               Pointer to MADT table
+  @retval     EFI_SUCCESS           Patched the MADT
+              EFI_UNSUPPORTED       Sysinfo table not found.
+              EFI_OUT_OF_RESOURCES  If less MADT entries
+**/
+EFI_STATUS
+UpdateFwst (
+  IN UINT8   *Current
+  )
+{
+  UINT32        RsvdBase;
+  UINT32        EsrtTableSize;
+  UINT64        OemId;
+  EFI_STATUS    Status;
+  LOADER_GLOBAL_DATA              *LdrGlobal;
+  BOOT_LOADER_VERSION             *Version;
+  EFI_SYSTEM_RESOURCE_TABLE       *pEsrtTable;
+  EFI_SYSTEM_RESOURCE_ENTRY       *pEsrtEntry;
+  FIRMWARE_UPDATE_STATUS          *FwUpdateStatus;
+  EFI_FWST_ACPI_DESCRIPTION_TABLE *pFwstAcpiTable;
+
+  LdrGlobal = (LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer();
+
+  //
+  // We do not need to populate ACPI table during firmware update
+  //
+  if (LdrGlobal->BootMode == BOOT_ON_FLASH_UPDATE) {
+    return EFI_SUCCESS;
+  }
+
+  //
+  // Get bootloader reserved region base
+  //
+  Status = GetComponentInfo (FLASH_MAP_SIG_BLRESERVED, &RsvdBase, NULL);
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  //
+  // Allocate memory for ESRT structure
+  //
+  EsrtTableSize = sizeof (EFI_SYSTEM_RESOURCE_TABLE_HEADER) + sizeof (EFI_SYSTEM_RESOURCE_ENTRY);
+  pEsrtTable = AllocatePages (EFI_SIZE_TO_PAGES (EsrtTableSize));
```

Comment:
Why don't we just use ACPI memory instead ?

Reply:
Modified code to use ACPI memory